tools/rpp/src/lib.rs

```diff
@@ -0,0 +1,206 @@
+use std::collections::HashMap;
+use std::fmt;
+use std::fs;
+
+use regex::Regex;
+
+#[derive(Debug, Clone)]
+pub struct ParsingError {
+    message: String,
+    line: u32,
+}
+
+impl fmt::Display for ParsingError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.message)
+    }
+}
+
+impl ParsingError {
+    pub fn new(message: String, line: u32) -> ParsingError {
+        ParsingError {
+            message: message,
+            line: line,
+        }
+    }
+}
+
+struct Keyword {
+    name: &'static str,
+    process: fn(&str, &mut Context) -> Result<String, ParsingError>,
+}
+
+pub struct Context {
+    macros: HashMap<String, String>,
+    line_num: u32,
+
+    // Compiling regex is slow. Only do it once per context
+    word_re: Regex,
+    block_comment_re: Regex,
+    eol_comment_re: Regex,
+}
+
+impl Context {
+    pub fn new() -> Context {
+        Context {
+            macros: HashMap::new(),
+            line_num: 0,
+
+            word_re: Regex::new(r"\b\w+\b").unwrap(),
+            block_comment_re: Regex::new(r"(?s)/\*.*?\*/").unwrap(),
+            eol_comment_re: Regex::new(r"//.*$").unwrap(),
+        }
+    }
+}
+
+const KEYWORDS: &[Keyword] = &[
+    Keyword {
+        name: "#define",
+        process: process_define,
+    },
+    Keyword {
+        name: "#include",
+        process: process_include,
+    },
+];
+
+fn process_define(value: &str, ctx: &mut Context) -> Result<String, ParsingError> {
+    let mut parts = value.split_whitespace();
+    let key = match parts.next() {
+        Some(k) => k,
+        None => {
+            return Err(ParsingError::new(
+                String::from("#define has no name"),
+                ctx.line_num,
+            ))
+        }
+    };
```

Comment:
Looks like this could be written a bit shorter like this:
```rust
    let key = parts.next().ok_or_else(|| ParsingError::new(...))?;
```
Similarly in other places in this PR.

Reply:
Thanks for the suggestion! Done.